src/game/server/entities/character.cpp

```diff
@@ -2831,12 +2856,10 @@ bool CCharacter::TakeDamage(vec2 Force, int Dmg, int From, int Weapon, int Mode)
 	// m_pPlayer only inflicts half damage on self
 	if(From == m_pPlayer->GetCID())
 	{
-		if(GetClass() == PLAYERCLASS_HERO || (GetClass() == PLAYERCLASS_SOLDIER && m_ActiveWeapon == WEAPON_GRENADE)
-										  || (GetClass() == PLAYERCLASS_SCIENTIST && Mode == TAKEDAMAGEMODE_NOINFECTION)
-										  || (GetClass() == PLAYERCLASS_LOOPER && m_ActiveWeapon == WEAPON_GRENADE))
-			return false; // no self harm
-		else
+		if (Mode == TAKEDAMAGEMODE_SELFHARM)
```

Comment:
Self damage was not programmed very well. For example you are hero shoot a grenade up in the air, change weapon to laser rifle and when the grenade hits you you take damage. (while you should not) This change makes the code cleaner more flexible and fixes bugs. It is also necessary for the turrets, otherwise they would harm the owner. Merc will still be self harmed by its bombs. The same goes for scientist and soldier. It should not change the amount of selfharm the scientist receives.